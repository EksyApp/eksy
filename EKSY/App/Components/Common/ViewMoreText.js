import React, {Component} from 'react'
import { Text, View } from 'react-native'

const emptyFunc = ()=>{};

export class ViewMoreText extends Component {
  constructor(props){
    super(props)
    this.resetData()
    this.state = {
    numberOfLines: null,
    opacity: 0
    }
  }

  isTruncated = false
  originalHeight =  0
  shouldShowMore = false
  contentHeight = 0
  isInit = false

resetData(){
  this.isTruncated = false;
  this.originalHeight = 0;
  this.shouldShowMore = false;
  this.isInit = false;
}
  onLayout(event){
    const {x, y, width, height} = event.nativeEvent.layout;

    if(height === 0 || this.state.opacity === 1) return false;

    this.setOriginalHeight(height);
    this.checkTextTruncated(height);
    if(this.state.numberOfLines === this.props.numberOfLines){
      this.setState({
        opacity: 1
      })
    }
  }

  setOriginalHeight(height){
    if(this.originalHeight === 0){
      this.originalHeight = height;

      this.setState({
        numberOfLines: this.props.numberOfLines
      })
    }
  }

  checkTextTruncated(height){
    if(height < this.originalHeight){
      this.shouldShowMore = true;
    }
  }

  onPressMore(){
    (this.props.afterExpand || emptyFunc)();
    this.setState({
      numberOfLines: null
    });
  }

  onPressLess(){
    (this.props.afterCollapse || emptyFunc)();
    this.setState({
      numberOfLines: this.props.numberOfLines
    })
  }

  renderViewMore(){
    return (
      <Text onPress={() => {this.onPressMore()}}>
        View More
      </Text>
    )
  }

  renderViewLess(){
    return (
      <Text onPress={() => {this.onPressLess()}}>
        View Less
      </Text>
    )
  }

  renderFooter(){
    let {
      numberOfLines
    } = this.state;

    if (this.shouldShowMore === true){
      if(numberOfLines > 0) {
        return (this.props.renderViewMore || this.renderViewMore.bind(this))(this.onPressMore)
      } else {
        return (this.props.renderViewLess || this.renderViewLess.bind(this))(this.onPressLess)
      }
    }
  }

  render(){
    return (
      <View onLayout={(event) => {this.onLayout(event)}} style={{opacity: this.state.opacity}}>
        <Text
          numberOfLines={this.state.numberOfLines}>
          {this.props.children}
        </Text>
        {this.renderFooter()}

        {
          this.state.numberOfLines &&
          <View style={{width: 1, height: 1}}></View>
        }
      </View>
    )
  }
}

ViewMoreText.propTypes = {
  renderViewMore: React.PropTypes.func,
  renderViewLess: React.PropTypes.func,
  afterCollapse: React.PropTypes.func,
  afterExpand: React.PropTypes.func,
  numberOfLines: React.PropTypes.number.isRequired
}
