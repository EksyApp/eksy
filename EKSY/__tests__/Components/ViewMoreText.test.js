import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import ViewMoreText from "../../App/Components/Common/ViewMoreText";

test('ViewMoreText renders correctly', () => {
	const tree = renderer.create(
			<ViewMoreText />
	).toJSON();
	expect(tree).toMatchSnapshot();
});