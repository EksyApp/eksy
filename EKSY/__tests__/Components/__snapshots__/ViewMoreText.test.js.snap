// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewMoreText renders correctly when numberOfLines not null 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "opacity": 0,
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    numberOfLines={null}
    style={
      Array [
        Object {
          "color": "rgb(89,85,78)",
          "fontSize": 18,
          "margin": 10,
        },
        undefined,
      ]
    }
  />
</View>
`;

exports[`ViewMoreText renders correctly when numberOfLines null 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "opacity": 0,
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    numberOfLines={null}
    style={
      Array [
        Object {
          "color": "rgb(89,85,78)",
          "fontSize": 18,
          "margin": 10,
        },
        undefined,
      ]
    }
  />
</View>
`;
